File: 0955-delete-columns-to-make-sorted-ii/0955-delete-columns-to-make-sorted-ii.cpp
class Solution {
public:
    int minDeletionSize(vector<string>& strs) {
        int res = 0;
        int start = 0;
        int violation = false;
        vector<bool> sorted(strs.size() - 1, false);
        int unSorted = strs.size() - 1;
        for (int i = 0; i < strs[0].length() && unSorted > 0; i++) {
            violation = false;
            for (int j = 1; j < strs.size(); j++) {
                if (!sorted[j - 1] && strs[j][i] < strs[j - 1][i]) {
                    res++;
                    violation = true;
                    break;
                }
            }
            if(!violation)
            {
                for(int j = 1; j < strs.size(); j++)
                {
                    if(!sorted[j - 1] && strs[j][i] > strs[j - 1][i])
                    {
                        sorted[j - 1] = true;
                        unSorted--;
                    }
                }
            }
        }
        return res;
    }
};

File: 0955-delete-columns-to-make-sorted-ii/0955-delete-columns-to-make-sorted-ii.py
class Solution:
    def minDeletionSize(self, strs: List[str]) -> int:
        res = 0
        violation = False
        sortedIdxs = [False for _ in range(len(strs) - 1)]
        unSorted = len(strs) - 1
        i = 0
        while i < len(strs[0]) and unSorted > 0:
            violation = False
            for j in range(1, len(strs)):
                if not sortedIdxs[j - 1] and ord(strs[j][i]) < ord(strs[j - 1][i]):
                    res += 1
                    violation = True
                    break
            if not violation:
                for j in range(1, len(strs)):
                    if not sortedIdxs[j - 1] and ord(strs[j][i]) > ord(strs[j - 1][i]):
                        sortedIdxs[j - 1] = True
                        unSorted -= 1
            i += 1
        return res
